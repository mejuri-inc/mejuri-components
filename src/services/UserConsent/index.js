
const CONSENT_CATEGORIES_DICTIONARY = {
  'C0001' : 'Strictly Necessary',
  'C0002' : 'Performance',
  'C0003' : 'Functional',
  'C0004' : 'Targeting',
  'C0005' : 'Social Media'
}

/*
  Map services with the cookie categories that corresponds
*/
const SERVICES_CATEGORIES = {
  'Sailthru'                :['Targeting','Functional','Performance'],
  'Google Analytics'        :['Performance'],
  'Pinterest Tag'           :['Targeting'],
  'Facebook Pixel'          :['Targeting'],
  'Twitter Ads'             :['Targeting'],
  'Google Tag Manager'      :['Strictly Necessary','Targeting','Functional','Performance','Social Media'],
  'Mouseflow'               :['Performance'],
  'Vimeo'                   :['Performance'],
  'AdRoll'                  :['Performance','Targeting'],
  'Google Ads (Gtag)'       :['Performance','Targeting'],
  'Podsights'               :['Performance','Targeting'],
  'Heap'                    :['Performance'],
  'ShareASale'              :['Performance','Targeting'],
  'The Trade Desk (Mejuri)' :['Performance','Targeting'],
  'Privy'                   :['Performance','Functional']
}


class UserConsent{


  // IF THIS FLAG IS PRESENT ALWAYS ENABLE ALL CATEGORIES
  BYPASS_PRIVACY_PREFERENCES = false

  consentCategories   = []
  activeCategories    = []
  servicesList        = []
  availableServices   = []
  unavailableServices = []
  callbacks           = []
  scripts             = []

  constructor(){
    this.onUpdate()
    if(typeof window !== 'undefined'){
      window.addEventListener('consent.onetrust',this._onConsentChange.bind(this))
      window.addEventListener('load',this._init.bind(this))
    }
  }

  onConsentChange(callback){
    this.callbacks.push(callback)
  }

  removeOnConsentChange(callback){
      this.callbacks = this.callbacks.filter( c => callback != c )
  }

  _onConsentChange(e){
    this.onUpdate()
    this.callbacks.forEach( cb => cb(this,e))
  }

  onUpdate(){
    this.consentCategories    = this.getConsentCategories()
    this.activeCategories     = this.getActiveCategories()
    this.servicesList         = this.getServicesList()
    this.availableServices    = this.getAvailableServices()
    this.unavailableServices  = this.getUnavailableServices()
    this._injectCode()
  }

  getActiveCategories(){
    return this.OtActiveGroups.map( id => CONSENT_CATEGORIES_DICTIONARY[id] || id )
  }

  get OtActiveGroups(){
    let activeGroups = [], OnetrustActiveGroups
    if(typeof window !== 'undefined') OnetrustActiveGroups = window.OnetrustActiveGroups
    if(this.BYPASS_PRIVACY_PREFERENCES){
      activeGroups = Object.keys(CONSENT_CATEGORIES_DICTIONARY)
    }else if(OnetrustActiveGroups){
      activeGroups = OnetrustActiveGroups.split(",").filter( id => id )
    }
    return activeGroups
  }

  getConsentCategories(){
    let categories = CONSENT_CATEGORIES_DICTIONARY
    return Object.keys(categories).map(key => categories[key])
  }

  isServiceAvailable(service){
    let available = false
    for(let category of SERVICES_CATEGORIES[service]){
      available = this.activeCategories.includes(category)
      if(!available) return
    }
    return available
  }

  getServicesList(){
    return Object.keys(SERVICES_CATEGORIES)
  }

  getServiceCategoriesId(service){
    const dictionary = CONSENT_CATEGORIES_DICTIONARY
    const serviceCategories = SERVICES_CATEGORIES[service]
    return serviceCategories.map(
      categoryName => {
        let categoryId
        for(let id in dictionary){
          if(dictionary[id] == categoryName){
            return categoryId = id
          }
        }
        return categoryId
      }
    )
  }

  getAvailableServices(){
    return this.servicesList.filter(this.isServiceAvailable.bind(this))
  }

  getUnavailableServices(){
    return this.servicesList.filter((service)=> !this.isServiceAvailable(service))
  }

  registerScript(service, code, parentNode){
    this._registerScript({ service, code, parentNode })
    /*
      Right after the script is registered,
      we'll try to inject it in case user already gave consent
    */
    this._injectCode()
  }

  _registerScript(service, code, parentNode = 'head'){
    this.scripts.push({ service, code, parentNode })
  }

  insertScriptTag({ code , parentNode }){
    const tag = document.createElement('script')
    tag.type  = 'text/javascript'
    document.querySelector(parentNode).appendChild(tag)
    tag.src   = code
  }

  _injectCode(){
    this.scripts.filter( script =>{
      let keep = true
      if(this.availableServices.includes(script.service)){
        switch(typeof script.code){
          case 'function':
            script.code()
            keep = false
            break
          case 'string':
            this.insertScriptTag(script)
            keep = false
            break
        }
      }
      return keep
    })
  }

  _init(){
    if(this.OtActiveGroups.length < 1 && !this.BYPASS_PRIVACY_PREFERENCES){
      setTimeout(this._init.bind(self),500)
    }else{
      this._onConsentChange()
    }
  }

}

const singleton = new UserConsent()

export default singleton
