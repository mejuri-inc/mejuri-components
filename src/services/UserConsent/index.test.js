import UserConsent from './index'

describe('UserConsent', () => {

  beforeEach(()=>{
    UserConsent.BYPASS_PRIVACY_PREFERENCES = false
  })

  //-------------------------------------------------------------------------------

  it(`Available services match user preferences (Strictly necessary)`, async () => {
    window.OnetrustActiveGroups = ',C0001,'
    UserConsent.onUpdate()
    expect(UserConsent.getAvailableServices()).toMatchSnapshot()
  })

  it(`Active categories match user preferences (Strictly necessary)`, async () => {
    window.OnetrustActiveGroups = ',C0001,'
    UserConsent.onUpdate()
    expect(UserConsent.getActiveCategories()).toMatchSnapshot()
  })

  //-------------------------------------------------------------------------------

  it(`Available services match user preferences (Performance)`, async () => {
    window.OnetrustActiveGroups = ',C0002,C0001,'
    UserConsent.onUpdate()
    expect(UserConsent.getAvailableServices()).toMatchSnapshot()
  })

  it(`Active categories match user preferences (Performance)`, async () => {
    window.OnetrustActiveGroups = ',C0002,C0001,'
    UserConsent.onUpdate()
    expect(UserConsent.getActiveCategories()).toMatchSnapshot()
  })

  //-------------------------------------------------------------------------------

  it(`Available services match user preferences (Functional)`, async () => {
    window.OnetrustActiveGroups = ',C0003,C0001,'
    UserConsent.onUpdate()
    expect(UserConsent.getAvailableServices()).toMatchSnapshot()
  })

  it(`Active categories match user preferences (Functional)`, async () => {
    window.OnetrustActiveGroups = ',C0003,C0001,'
    UserConsent.onUpdate()
    expect(UserConsent.getActiveCategories()).toMatchSnapshot()
  })

  //-------------------------------------------------------------------------------

  it(`Available services match user preferences (Targeting)`, async () => {
    window.OnetrustActiveGroups = ',C0004,C0001,'
    UserConsent.onUpdate()
    expect(UserConsent.getAvailableServices()).toMatchSnapshot()
  })

  it(`Active categories match user preferences (Targeting)`, async () => {
    window.OnetrustActiveGroups = ',C0004,C0001,'
    UserConsent.onUpdate()
    expect(UserConsent.getActiveCategories()).toMatchSnapshot()
  })

  //-------------------------------------------------------------------------------

  it(`Available services match user preferences (Social Media)`, async () => {
    window.OnetrustActiveGroups = ',C0005,C0001,'
    UserConsent.onUpdate()
    expect(UserConsent.getAvailableServices()).toMatchSnapshot()
  })

  it(`Active categories match user preferences (Social Media)`, async () => {
    window.OnetrustActiveGroups = ',C0005,C0001,'
    UserConsent.onUpdate()
    expect(UserConsent.getActiveCategories()).toMatchSnapshot()
  })

  //-------------------------------------------------------------------------------

  it(`All services are avialable when the bypass option is enabled`, async () => {
    window.OnetrustActiveGroups = ',C0001,'
    UserConsent.BYPASS_PRIVACY_PREFERENCES = true
    UserConsent.onUpdate()
    expect(UserConsent.getAvailableServices()).toMatchSnapshot()
  })

  it(`All categories are active when the bypass option is enabled`, async () => {
    window.OnetrustActiveGroups = ',C0001,'
    UserConsent.BYPASS_PRIVACY_PREFERENCES = true
    UserConsent.onUpdate()
    expect(UserConsent.getActiveCategories()).toMatchSnapshot()
  })





})
