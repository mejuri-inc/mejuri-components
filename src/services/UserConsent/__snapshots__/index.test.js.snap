// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserConsent Active categories match user preferences (Functional) 1`] = `
Array [
  "Functional",
  "Strictly Necessary",
]
`;

exports[`UserConsent Active categories match user preferences (Performance) 1`] = `
Array [
  "Performance",
  "Strictly Necessary",
]
`;

exports[`UserConsent Active categories match user preferences (Social Media) 1`] = `
Array [
  "Social Media",
  "Strictly Necessary",
]
`;

exports[`UserConsent Active categories match user preferences (Strictly necessary) 1`] = `
Array [
  "Strictly Necessary",
]
`;

exports[`UserConsent Active categories match user preferences (Targeting) 1`] = `
Array [
  "Targeting",
  "Strictly Necessary",
]
`;

exports[`UserConsent Available services match user preferences (Functional) 1`] = `Array []`;

exports[`UserConsent Available services match user preferences (Performance) 1`] = `
Array [
  "Google Analytics",
  "Mouseflow",
  "Vimeo",
  "Heap",
]
`;

exports[`UserConsent Available services match user preferences (Social Media) 1`] = `Array []`;

exports[`UserConsent Available services match user preferences (Strictly necessary) 1`] = `Array []`;

exports[`UserConsent Available services match user preferences (Targeting) 1`] = `
Array [
  "Pinterest Tag",
  "Facebook Pixel",
  "Twitter Ads",
]
`;
